core/src/repair/serve_repair.rs

```diff
@@ -215,9 +215,12 @@ impl RepairRequestHeader {
 
 pub(crate) type Ping = ping_pong::Ping<[u8; REPAIR_PING_TOKEN_SIZE]>;
 
+// 142 bytes when bincode serialized to match gossip::LegacyContactInfo.
+type LegacyContactInfo = ([u64; 17], [u8; 6]);
```

Comment:
may be a dumb question. But how do we know `LegacyContactInfo` is always 142 bytes? Can't `SockerAddr` vary in length depending if it is V4 or V6?

Reply:
oh i guess Enum values are stored at the size of largest discriminant, so size of V6. although SocketAddrV6 size varies on OS: https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html. Wondering if we'll run into an issue here?

Reply:
> But how do we know LegacyContactInfo is always 142 bytes?
> Can't SockerAddr vary in length depending if it is V4 or V6?

yeah, it can depend on `SocketAddr` enum variant.
I just used `LegacyContactInfo::default()` which creates ipv4.

> oh i guess Enum values are stored at the size of largest discriminant, so size of V6.

That is only for the in-memory layout of enums.
For `bincode::serialize`, the serialized size indeed depends on the actual enum variant 

-------------------------

Thinking more about the code here, I don't think we need to add this type to begin with. We can just remove all fields (including legacy contact-info) from `RepairProtocol::Legacy...` variants.
This will preserve the enum identifier for the serialize `RepairProtocol` for backward compatibility.
Additionally if someone sends a `RepairProtocol::Legacy...` it will fail in deserialization because of `reject_trailing_bytes`:
https://github.com/anza-xyz/agave/blob/075c79039/core/src/repair/serve_repair.rs#L1396
https://github.com/anza-xyz/agave/blob/075c79039/sdk/src/packet.rs#L177

(even without `reject_trailing_bytes` these requests are still skipped by the pipeline so we are good).

Reply:
A few followup questions, mostly just clarifying your comments:

>  We can just remove all fields (including legacy contact-info) from RepairProtocol::Legacy... variants.
This will preserve the enum identifier for the serialize RepairProtocol for backward compatibility.

Since this enum is NOT local to a node (ie we send in packets between nodes), we want to keep the legacy enum variants to avoid misidentification of packets. It looks like the byte lengths wouldn't match in our case so we technically may not be at risk, but in general, leaving the legacy values the right thing to do. Is this all correct ?

> Additionally if someone sends a RepairProtocol::Legacy... it will fail in deserialization because of reject_trailing_bytes:

Makes sense to me - and if deserialize fails, we'll just ignore and increment an error counter
https://github.com/anza-xyz/agave/blob/32a89558014d62cb6928c233bd010c27dba34171/core/src/repair/serve_repair.rs#L520-L521

Reply:
> Since this enum is NOT local to a node (ie we send in packets between nodes), we want to keep the legacy enum variants to avoid misidentification of packets. It looks like the byte lengths wouldn't match in our case so we technically may not be at risk, but in general, leaving the legacy values the right thing to do. Is this all correct ?

yes

> By this line of thinking, I don't think Firedancer would care about this change (in the sense that it'd impact them), but maybe we still give them a heads up ? 

No objection to giving a heads up, but this change shouldn't impact any other client anyways: We are not changing wire-format of _outgoing_ packets (`Legacy...` ones are never sent out). Valid packets are the same as before.
And for invalid _incoming_ packets, whether or not they deserialize as a `Legacy...` request, they are skipped/discarded anyways. 

> Do you happen to know if they've implemented repair, and if they did, if they did only the new requests or if they did legacy too ?

They have implemented repair, but I have not looked up their implementation detail. (again, I don't think it matters though).

Reply:
> No objection to giving a heads up, but this change shouldn't impact any other client anyways: We are not changing wire-format of outgoing packets (Legacy... ones are never sent out). Valid packets are the same as before.
> And for invalid incoming packets, whether or not they deserialize as a Legacy... request, they are skipped/discarded anyways.

Awesome, these were the items I was looking for confirmation on